### PR TITLE
Delete implicit conversion from StringLiteral to std::string.

### PR DESCRIPTION
**Pull request by BillyONeal**
Initially an attempted resolution of https://github.com/microsoft/vcpkg/issues/21142 , we've decided to keep even though it ended up being unrelated.

### CODE REVIEW

File: include/vcpkg/base/json.h [hunk 1]
```diff
@@ -203,22 +209,35 @@ namespace vcpkg::Json
...
-        Object& insert(std::string key, const Object& value);
-        Array& insert(std::string key, Array&& value);
-        Array& insert(std::string key, const Array& value);
+        template<class StringViewLike,
+                 class = std::enable_if_t<std::is_constructible<StringView, StringViewLike&&>::value>>
+        Value& insert(StringView key, StringViewLike&& value)
+        {
+            return this->insert(key, StringView(static_cast<StringViewLike&&>(value)).to_string());
+        }
+        Value& insert(StringView key, std::string&& value);
```
Comment (ras0219-msft) [inline include/vcpkg/base/json.h:218]: @BillyONeal Calling explicit attention to these lines as ones I adjusted.

File: include/vcpkg/base/json.h [hunk 2]
```diff
@@ -203,22 +209,35 @@ namespace vcpkg::Json
...
-        Object& insert_or_replace(std::string key, const Object& value);
-        Array& insert_or_replace(std::string key, Array&& value);
-        Array& insert_or_replace(std::string key, const Array& value);
+        template<class StringViewLike,
+                 class = std::enable_if_t<std::is_constructible<StringView, StringViewLike&&>::value>>
+        Value& insert_or_replace(StringView key, StringViewLike&& value)
+        {
+            return this->insert_or_replace(key, StringView(static_cast<StringViewLike&&>(value)).to_string());
+        }
+        Value& insert_or_replace(StringView key, std::string&& value);
```
Comment (ras0219-msft) [inline include/vcpkg/base/json.h:234]: @BillyONeal  Needs explicit review

**Review by ras0219-msft — APPROVED** [review]
@BillyONeal if my additional commit looks good to you, this is ready to merge

File: include/vcpkg/base/json.h (hunk 1 above)
Comment (BillyONeal) [reply include/vcpkg/base/json.h]: Proposal:

```
        template<class StringViewLike,
                 class = std::enable_if_t<std::is_constructible<StringView, const StringViewLike&>::value>>
        Value& insert(StringView key, const StringViewLike& value)
```

We are never moving from this input so we can use constT& which is much less psycho greedy.

File: include/vcpkg/base/json.h [hunk 3]
```diff
@@ -120,7 +120,13 @@ namespace vcpkg::Json
         static Value boolean(bool) noexcept;
         static Value integer(int64_t i) noexcept;
         static Value number(double d) noexcept;
-        static Value string(std::string s) noexcept;
+        template<class StringLike>
+        static Value string(StringLike&& s) noexcept
+        {
+            return string(StringView(s).to_string());
+        }
+        static Value string(std::string&& s) noexcept;
```
Comment (BillyONeal) [inline include/vcpkg/base/json.h]: I think we should reorder this one before the template (and also on 212). Yes I know C++ makes it not matter but I think it's clearer to read if the overload we intend to call is right before.

File: include/vcpkg/base/json.h (hunk 1 above)
Comment (BillyONeal) [reply include/vcpkg/base/json.h]: Also, I thought we couldn't use enable_if_t ... guess I should fix the others.

File: include/vcpkg/base/json.h [hunk 4]
```diff
@@ -209,11 +210,11 @@ namespace vcpkg::Json
...
         // asserts if the key is found
-        template<class StringViewLike,
-                 class = std::enable_if_t<std::is_constructible<StringView, StringViewLike&&>::value>>
-        Value& insert(StringView key, StringViewLike&& value)
+        template<class StringLike,
+                 std::enable_if_t<std::is_constructible<StringView, const StringLike&>::value, int> = 0>
+        Value& insert(StringView key, const StringLike& value)
         {
-            return this->insert(key, StringView(static_cast<StringViewLike&&>(value)).to_string());
+            return this->insert(key, StringView(value).to_string());
```
Comment (ras0219-msft) [inline include/vcpkg/base/json.h:217]: I believe this should still have a `static_cast<const StringLike&>` to capture string literal lengths.

File: include/vcpkg/base/json.h [hunk 5]
```diff
@@ -225,13 +226,13 @@ namespace vcpkg::Json
...
-        template<class StringViewLike,
-                 class = std::enable_if_t<std::is_constructible<StringView, StringViewLike&&>::value>>
-        Value& insert_or_replace(StringView key, StringViewLike&& value)
+        Value& insert_or_replace(StringView key, std::string&& value);
+        template<class StringLike,
+                 std::enable_if_t<std::is_constructible<StringView, const StringLike&>::value, int> = 0>
+        Value& insert_or_replace(StringView key, const StringLike& value)
         {
-            return this->insert_or_replace(key, StringView(static_cast<StringViewLike&&>(value)).to_string());
+            return this->insert_or_replace(key, StringView(value).to_string());
```
Comment (ras0219-msft) [inline include/vcpkg/base/json.h:234]: ditto above

File: include/vcpkg/base/messages.h [hunk 6]
```diff
@@ -32,17 +32,13 @@ namespace vcpkg
...
         {
-            if constexpr (std::is_same<StringLike, std::string>::value)
-            {
-                return LocalizedString(std::move(s));
-            }
-            else
-            {
-                return LocalizedString(StringView(s));
-            }
+            return LocalizedString(StringView(s));
```
Comment (ras0219-msft) [inline include/vcpkg/base/messages.h:41]: ditto above

**Review by ras0219-msft — APPROVED** [review]

File: include/vcpkg/base/json.h (hunk 4 above)
Comment (BillyONeal) [reply include/vcpkg/base/json.h:217]: That cast has no effect: https://gcc.godbolt.org/z/nfc6rqo3W

Comment (ras0219-msft) [reply include/vcpkg/base/json.h:217]: TIL, and am less happy as a result

**Pull request merged**